### PR TITLE
Implement local_obsdata as C++

### PR DESCRIPTION
**Pull request by joakim-hove**
On top of: https://github.com/equinor/ert/pull/2964

Implement `local_obsdata` as C++

### CODE REVIEW

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
I'll rebase this shortly - but coil if it can get a review

**Comment by joakim-hove** [comment]
test ert please

**Comment by joakim-hove** [comment]
test required please

**Comment by joakim-hove** [comment]
test required please

**Comment by joakim-hove** [comment]
test ert please

File: libres/lib/enkf/local_ministep.cpp [hunk 1]
```diff
@@ -156,6 +155,19 @@ void add_obsdata_node(py::handle self, LocalObsDataNode &node) {
...
+LocalObsData &get_obsdata(py::handle obj) {
+    auto *ministep = reinterpret_cast<local_ministep_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
+    auto *obsdata_ptr = local_ministep_get_obsdata(ministep);
+    return *obsdata_ptr;
+}
+
+void attach_obsdata(py::handle obj, LocalObsData &obsdata) {
+    auto *ministep = reinterpret_cast<local_ministep_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
```
Comment (oyvindeide) [inline libres/lib/enkf/local_ministep.cpp:167]: Use `from_cwrap`

File: libres/lib/enkf/local_ministep.cpp [hunk 2]
```diff
@@ -156,6 +155,19 @@ void add_obsdata_node(py::handle self, LocalObsDataNode &node) {
     local_ministep_add_obsdata_node(ministep, &node);
 }
 
+LocalObsData &get_obsdata(py::handle obj) {
+    auto *ministep = reinterpret_cast<local_ministep_type *>(
+        PyLong_AsVoidPtr(obj.attr("_BaseCClass__c_pointer").ptr()));
```
Comment (oyvindeide) [inline libres/lib/enkf/local_ministep.cpp:160]: Use `from_cwrap`

File: res/enkf/local_config.py [hunk 3]
```diff
@@ -136,7 +124,7 @@ def getMinistep(self, mini_step_key):
     def getObsdata(self, obsdata_key):
```
Comment (oyvindeide) [inline res/enkf/local_config.py:124]: As far as I can tell this functionality is never called in the tests? Can you add a test to make sure the behavior stays the same?

File: res/enkf/local_obsdata.py [hunk 4]
```diff
@@ -1,138 +1,3 @@
...
-    def free(self):
-        self._free()
-
-    def __repr__(self):
-        return "LocalObsdata(len = %d, name = %s) at 0x%x" % (
-            len(self),
-            self.name(),
-            self._address(),
-        )
+__all__ = ["LocalObsdata"]
```
Comment (oyvindeide) [inline res/enkf/local_obsdata.py:3]: Think this has a few integration test, but very few unit tests, would be good to add some to make sure the behavior does not change.

File: tests/libres_tests/res/enkf/test_local_config.py [hunk 5]
```diff
@@ -119,14 +119,14 @@ def test_all_active(self):
...
-            self.assertEqual(1, ministep.numActiveData())
-            self.assertTrue(ministep.hasActiveData("PERLIN_PARAM"))
+            # ministep = updatestep[0]
+            # self.assertEqual(1, ministep.numActiveData())
+            # self.assertTrue(ministep.hasActiveData("PERLIN_PARAM"))
 
-            obsdata = ministep.getLocalObsData()
-            self.assertEqual(len(obsdata), 3)
+            # obsdata = ministep.getLocalObsData()
+            # self.assertEqual(len(obsdata), 3)
```
Comment (oyvindeide) [inline tests/libres_tests/res/enkf/test_local_config.py:129]: Is this no longer valid?

File: tests/libres_tests/res/enkf/test_local_config.py [hunk 6]
```diff
@@ -75,7 +75,7 @@ def test_local_obs_data(self):
             self.assertTrue(node_added)
 
             # Error when adding existing obs node
-            with self.assertRaises(KeyError):
+            with self.assertRaises(Exception):
```
Comment (oyvindeide) [inline tests/libres_tests/res/enkf/test_local_config.py:78]: Is this a base `Exception` or a specific one?

**Review by oyvindeide — COMMENTED** [review]
Looks like a good change, some comments :+1:

**Comment by joakim-hove** [comment]
Thanks for review comments - unfortunately offline for sine time, but will adress asap.

File: libres/lib/enkf/local_ministep.cpp (hunk 1 above)
Comment (joakim-hove) [reply libres/lib/enkf/local_ministep.cpp:167]: OK

File: libres/lib/enkf/local_ministep.cpp (hunk 2 above)
Comment (joakim-hove) [reply libres/lib/enkf/local_ministep.cpp:160]: ok

File: tests/libres_tests/res/enkf/test_local_config.py [hunk 7]
```diff
@@ -119,14 +119,14 @@ def test_all_active(self):
             main = test_context.getErt()
 
             local_config = main.getLocalConfig()
-
+            print("Have returned local config")
```
[the remark this answers is not in the document]
Comment (joakim-hove) [reply tests/libres_tests/res/enkf/test_local_config.py:122]: ok

File: tests/libres_tests/res/enkf/test_local_config.py (hunk 6 above)
Comment (joakim-hove) [reply tests/libres_tests/res/enkf/test_local_config.py:78]: Should be `KeyError` - did not know how to throw that from C++; got a tip from Zohar.

File: tests/libres_tests/res/enkf/test_local_config.py (hunk 5 above)
Comment (joakim-hove) [reply tests/libres_tests/res/enkf/test_local_config.py:129]: Hmmm - no that should still be valid; have reverted the diff.

File: res/enkf/local_obsdata.py (hunk 4 above)
Comment (joakim-hove) [reply res/enkf/local_obsdata.py:3]: OK: this is tested here: https://github.com/equinor/ert/pull/3046

File: res/enkf/local_config.py (hunk 3 above)
Comment (joakim-hove) [reply res/enkf/local_config.py:124]: Tested here: #3046

**Comment by joakim-hove** [comment]
OK: makes sense to review and merge #3041 first.

**Comment by joakim-hove** [comment]
OK - the tests requested have now been merged in with #3046

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

**Comment by joakim-hove** [comment]
> Can one of the admins verify this patch?

That would be cool :grin:

**Comment by joakim-hove** [comment]
ping :shrug:

**Review by oyvindeide — APPROVED** [review]
LGTM!

**Pull request merged**